Objects/unicodeobject.c

```diff
@@ -4900,13 +4892,32 @@ unicode_decode_utf8(const char *s, Py_ssize_t size,
         return get_latin1_char((unsigned char)s[0]);
     }
 
+    const char *starts = s;
+    const char *end = s + size;
+
+    // fast path: try ASCII string.
+    PyObject *u = PyUnicode_New(size, 127);
+    if (u == NULL) {
+        return NULL;
+    }
+    s += ascii_decode(s, end, PyUnicode_DATA(u));
+    if (s == end) {
+        return u;
+    }
+
+    // Use _PyUnicodeWriter after fast path is failed.
+    _PyUnicodeWriter writer;
     _PyUnicodeWriter_Init(&writer);
     writer.min_length = size;
-    if (_PyUnicodeWriter_Prepare(&writer, writer.min_length, 127) == -1)
-        goto onError;
+    writer.buffer = u;
+    writer.pos = s - starts;
+    _PyUnicodeWriter_Update(&writer);
```

Comment:
IMHO this pattern would deserve a new _PyUnicodeWriter_InitFromBuffer(writer, buffer) constructor.